src/arrow/schema.rs

```diff
@@ -0,0 +1,992 @@
+
+use std::collections::HashSet;
+use std::rc::Rc;
+
+use schema::types::SchemaDescPtr;
+use schema::types::Type;
+use schema::types::TypePtr;
+use basic::LogicalType;
+use basic::Type as PhysicalType;
+use basic::Repetition;
+
+use arrow_lib::datatypes::Schema;
+use arrow_lib::datatypes::Field;
+use arrow_lib::datatypes::DataType;
+
+use errors::ParquetError::ArrowError;
+use errors::Result;
+
+/// Convert parquet schema to arrow schema.
+pub fn parquet_to_arrow_schema(parquet_schema: SchemaDescPtr) -> Result<Schema> {
+  parquet_to_arrow_schema_by_columns(parquet_schema.clone(),
+                                     0..parquet_schema.columns().len())
+}
+
+/// Convert parquet schema to arrow schema, only preserving some leaf columns.
+pub fn parquet_to_arrow_schema_by_columns<T>(
+  parquet_schema: SchemaDescPtr, column_indices: T) -> Result<Schema>
+  where T: IntoIterator<Item=usize> {
+  let mut base_nodes = Vec::new();
+  let mut base_nodes_set = HashSet::new();
+  let mut leaves = HashSet::new();
+  
+  for c in column_indices {
+    let column = parquet_schema.column(c).self_type() as *const Type;
+    let root = parquet_schema.get_column_root_ptr(c);
+    let root_raw_ptr = root.clone().as_ref() as *const Type;
+    
+    leaves.insert(column);
+    if !base_nodes_set.contains(&root_raw_ptr) {
+      base_nodes.push(root);
+      base_nodes_set.insert(root_raw_ptr);
+    }
+  }
+  
+  let leaves = Rc::new(leaves);
+  base_nodes.into_iter()
+    .map(|t| ParquetTypeConverter::new(t, leaves.clone()).to_field())
+    .collect::<Result<Vec<Option<Field>>>>()
+    .map(|result| result.into_iter()
+      .filter_map(|f| f)
+      .collect::<Vec<Field>>())
+    .map(|fields| Schema::new(fields))
+}
+
+/// This struct is used to group methods and data structures used to convert parquet
+/// schema together.
+struct ParquetTypeConverter {
+  schema: TypePtr,
+  leaves: Rc<HashSet<*const Type>>
+}
+
+impl ParquetTypeConverter {
+  fn new(schema: TypePtr, leaves: Rc<HashSet<*const Type>>) -> Self {
+    Self {
+      schema,
+      leaves
+    }
+  }
+  
+  fn clone_with_schema(&self, other: TypePtr) -> Self {
+    Self {
+      schema: other,
+      leaves: self.leaves.clone()
+    }
+  }
+}
+
+impl ParquetTypeConverter {
+  // Interfaces.
+  
+  fn to_data_type(&self) -> Result<Option<DataType>> {
+    match self.schema.as_ref() {
+      Type::PrimitiveType {..} => self.to_primitive_type(),
+      Type::GroupType {..} => self.to_group_type()
+    }
+  }
+  
+  fn to_field(&self) -> Result<Option<Field>> {
+    self.to_data_type().map(|opt| opt.map(|dt| {
+      Field::new(self.schema.name(), dt, self.is_nullable())
+    }))
+  }
+  
+  // Utility functions.
+  
+  fn is_nullable(&self) -> bool {
+    let basic_info = self.schema.get_basic_info();
+    if basic_info.has_repetition() {
+      match basic_info.repetition() {
+        Repetition::OPTIONAL => true,
+        Repetition::REPEATED => true,
```

Comment:
Is repeated field nullable? Normally it is complemented by nested optional or required fields. I am not very familiar with arrow, maybe it treats them as nullable by default.

Reply:
Yes, you're right. In fact there's no clear definition about the mapping to arrow and I just think this is default case.